code/datums/progress_bar.dm

```diff
@@ -0,0 +1,213 @@
+/datum/progress_bar
+	// Settings
+	var/name
+	/// The goal value to reach. When [/datum/progress_bar/proc/update] is called, a fraction is calculated with the goal.
+	var/goal = 100
+	/// The icon file to use for the visual element.
+	var/icon
+	/// The total height in pixels of the theme's progress bar.
+	var/height
+	/// The width in pixels of the theme's fill sprite.
+	var/fill_width
+	/// The total time that determines the speed for appear animation.
+	var/animation_time_appear = 0.5 SECOND
+	/// The total time that determines the speed for fade animation.
+	var/animation_time_fade = 0.5 SECOND
+	/// The total time that determines the speed for shift down animation.
+	var/animation_time_shift = 0.5 SECOND
+
+	// Variables
+	/// The target to display the progress bar on.
+	var/atom/target
+	/// The main visual element holding everything up.
+	var/image/holder
+	/// The backdrop visual element. Contains the fill and mask as overlays.
+	var/image/backdrop
+	/// The fill visual element.
+	var/image/fill
+	/// The mob currently owning the progress bar, used to calculate the bar's vertical position.
+	var/mob/user
+	/// The client currently owning the progress bar for display.
+	var/client/client
+	/// The bar's index number within the owning mob's active progress bar list.
+	var/list_index
+	/// Whether the bar is stopping.
+	var/stopping = FALSE
+	/// Stored progress number for the goal fail effect.
+	var/last_progress = 0
+
+/datum/progress_bar/New(mob/param_user, param_goal, atom/param_target)
+	. = ..()
+	if(!istype(param_target))
+		throw EXCEPTION("datum/progress_bar: Invalid param given - param_target")
+
+	target = param_target
+	user = param_user
+	goal = param_goal
+	init_images(target)
+
+	if(user)
+		client = user.client
+		client?.images |= holder
+
+		LAZYINITLIST(user.progress_bars)
+		LAZYINITLIST(user.progress_bars[target])
+		var/list/bars = user.progress_bars[target]
+		bars += src
+		list_index = length(bars)
+		animate(holder, pixel_z = world.icon_size + (height * (list_index - 1)), alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+	else
+		animate(holder, alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+
+/datum/progress_bar/Destroy()
+	stop()
```

Comment:
OK, so I read through this whole thing again, and this just doesn't pass muster with me. `stop` here has branching logic based on whether we are deleting or not; if you call stop (rather than qdel) it actually gets called twice, first to do cleanup and start the animation, and then again from here to do ???. On the other hand, if you call `qdel`stop is called once, from here, and the animation stuff is triggered but happens in nullspace.

This feels like a needlessly confusing way to structure the execution flow. Do the cleanup in one proc, do the animation in another. Have the call chain be (animation proc) -> (delay and call qdel) -> (Destroy calls the cleanup proc or just does the cleanup there). The cleanup probably shouldn't happen until the animation completes anyway. If I'm way off the mark with that, well, call the cleanup proc at the top of the animation proc too. You can then call the animation proc for a fancy deletion or qdel for a quick one.